src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java

```diff
@@ -100,6 +162,14 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
 
                 setOutputDone();
 
+                //now output rowset closed, so we can wait for our confirmation steps
+                if ( data.activeConfirmation )  {
+                    while (!isStopped() && data.watchedConfirmStep.size() > 0) {    
+                        Thread.sleep(500);

```

Comment:
i don't know exactly how step perfomed after finish ProcessRow and ClosedOutputStream.
Wait here is needed to be sure that control flow of this step will not moved futher toward flush or close anyway. and we have to wait for steps wich input row we are subscribed to to finish, to be sure that no message tags from them will not be missed.